diff --git a/DynmapCore/src/main/resources/extracted/web/js/chat.js b/DynmapCore/src/main/resources/extracted/web/js/chat.js
index e71f03ef..60fc95e2 100644
--- a/DynmapCore/src/main/resources/extracted/web/js/chat.js
+++ b/DynmapCore/src/main/resources/extracted/web/js/chat.js
@@ -4,13 +4,17 @@ componentconstructors['chat'] = function(dynmap, configuration) {
 	
 	if(dynmap.getBoolParameterByName("hidechat"))
 		return;
+
+	const seen = new Set();
 		
 	// Provides 'chat'-events by monitoring the world-updates.
 	$(dynmap).bind('worldupdate', function(event, update) {
 		swtch(update.type, {
 			chat: function() {
-				$(dynmap).trigger('chat', [{source: update.source, name: update.playerName, text: update.message, account: update.account,
-                channel: update.channel}]);
+				const event = {source: update.source, name: update.playerName, text: update.message, account: update.account, channel: update.channel};
+				const json = JSON.stringify(event);
+				if (!seen.has(json)) $(dynmap).trigger("chat", [event]);
+				seen.add(json);
 			}
 		});
 	});
@@ -26,7 +30,7 @@ componentconstructors['chat'] = function(dynmap, configuration) {
 			var data = '{"name":'+JSON.stringify(pname?pname:"")+',"message":'+JSON.stringify(message)+'}';
 			$.ajax({
 				type: 'POST',
-		        contentType: "application/json; charset=utf-8",
+                       contentType: "application/json; charset=utf-8",
 				url: config.url.sendmessage,
 				data: data,
 				dataType: 'json',
diff --git a/DynmapCore/src/main/resources/extracted/web/js/dynmaputils.js b/DynmapCore/src/main/resources/extracted/web/js/dynmaputils.js
index 936664aa..6af90c12 100644
--- a/DynmapCore/src/main/resources/extracted/web/js/dynmaputils.js
+++ b/DynmapCore/src/main/resources/extracted/web/js/dynmaputils.js
@@ -207,13 +207,18 @@ var DynmapTileLayer = L.TileLayer.extend({
 	},
 
 	_tickLoadQueue: function() {
-		if (this._loadingTiles.length > 4) {
+		if (this._loadingTiles.length > 12) {
 			return;
 		}
 
+		if (this._loadingTiles.length !== 0) {
+			$(dynmap).trigger("load-from-center");
+		}
+
 		var next = this._loadQueue.shift();
 
 		if (!next) {
+			$(dynmap).trigger("tile-queue-loaded");
 			return;
 		}
 
diff --git a/DynmapCore/src/main/resources/extracted/web/js/map.js b/DynmapCore/src/main/resources/extracted/web/js/map.js
index a0ec24b6..e76a94b1 100644
--- a/DynmapCore/src/main/resources/extracted/web/js/map.js
+++ b/DynmapCore/src/main/resources/extracted/web/js/map.js
@@ -1,5 +1,8 @@
 "use strict";
 //if (!console) console = { log: function() {} };
+if (window.dynUpdate == null) window.dynUpdate = function (url, cb) {
+	$.getJSON(url, cb);
+}
 
 var componentconstructors = {};
 var maptypes = {};
@@ -99,33 +102,23 @@ DynMap.prototype = {
 			});
 			me.defaultworld = me.defaultworld || world;
 		});
-		var urlarg = me.getParameterByName('worldname');
-		if(urlarg == "")
-			urlarg = me.options.defaultworld || "";
-		if(urlarg != "") {
-		    me.defaultworld = me.worlds[urlarg] || me.defaultworld;
-		}
-		urlarg = me.getParameterByName('mapname');
-		if(urlarg != "") {
-			me.defaultworld.defaultmap = me.defaultworld.maps[urlarg] || me.defaultworld.defaultmap;
-		}
-		urlarg = me.getIntParameterByName('x');
-		if(urlarg != null)
-			me.defaultworld.center.x = urlarg;
-		urlarg = me.getIntParameterByName('y');
-		if(urlarg != null)
-			me.defaultworld.center.y = urlarg;
-		urlarg = me.getIntParameterByName('z');
-		if(urlarg != null)
-			me.defaultworld.center.z = urlarg;
-		urlarg = me.getParameterByName('nogui');
-		if(urlarg != "") {
-			me.nogui = (urlarg == 'true');
-		}
-		urlarg = me.getParameterByName('nocompass');
-		if(urlarg != "") {
-			me.nocompass = (urlarg == 'true');
-		}
+
+		const parts = location.hash.slice(2).split(",");
+		const zoom = parseFloat(parts[0]);
+		const x = parseFloat(parts[1]);
+		const z = parseFloat(parts[2]);
+
+		if (isNaN(zoom) || isNaN(x) || isNaN(z)) return;
+
+		let world = parts[3];
+
+		if (world !== "world_nether" && world !== "world_the_end") world = "world";
+
+		me.defaultworld.defaultmap = me.defaultworld.maps[world] || me.defaultworld.defaultmap
+		me.defaultworld.center.x = x;
+		me.defaultworld.center.z = z;
+
+		me.options.defaultzoom = zoom;
 	},
 	initialize: function() {
 		var me = this;
@@ -620,7 +613,8 @@ DynMap.prototype = {
 		}
 
 		$(me).trigger('worldupdating');
-		$.getJSON(me.formatUrl('update', { world: me.world.name, timestamp: me.lasttimestamp, reqid: me.reqid }), function(update) {
+
+		window.dynUpdate(me.formatUrl('update', { world: me.world.name, timestamp: me.lasttimestamp, reqid: me.reqid }), function(update) {
 				me.reqid++; // Bump request ID always
 				if (!update) {
 					setTimeout(function() { me.update(); }, me.options.updaterate);
@@ -647,7 +641,6 @@ DynMap.prototype = {
 					me.lasttimestamp = update.timestamp;
 				}
 				if(me.options.confighash != update.confighash) {
-				    window.location = me.getLink();
 					return;
 				}
 				me.playerfield.text(me.options['msg-players'] + " [" + update.currentcount + "/" + me.options.maxcount + "]");
@@ -695,7 +688,9 @@ DynMap.prototype = {
 
 						swtch(update.type, {
 							tile: function() {
-								me.maptype.updateNamedTile(update.name, update.timestamp);
+								setTimeout(() => {
+									me.maptype.updateNamedTile(update.name, update.timestamp);
+								}, 1000);
 							},
 							playerjoin: function() {
 								$(me).trigger('playerjoin', [ update.playerName ]);
