diff --git a/DynmapCore/src/main/resources/extracted/web/js/chat.js b/DynmapCore/src/main/resources/extracted/web/js/chat.js
index 14a8501c..60fc95e2 100644
--- a/DynmapCore/src/main/resources/extracted/web/js/chat.js
+++ b/DynmapCore/src/main/resources/extracted/web/js/chat.js
@@ -4,13 +4,17 @@ componentconstructors['chat'] = function(dynmap, configuration) {
 	
 	if(dynmap.getBoolParameterByName("hidechat"))
 		return;
+
+	const seen = new Set();
 		
 	// Provides 'chat'-events by monitoring the world-updates.
 	$(dynmap).bind('worldupdate', function(event, update) {
 		swtch(update.type, {
 			chat: function() {
-				$(dynmap).trigger('chat', [{source: update.source, name: update.playerName, text: update.message, account: update.account,
-                channel: update.channel}]);
+				const event = {source: update.source, name: update.playerName, text: update.message, account: update.account, channel: update.channel};
+				const json = JSON.stringify(event);
+				if (!seen.has(json)) $(dynmap).trigger("chat", [event]);
+				seen.add(json);
 			}
 		});
 	});
diff --git a/DynmapCore/src/main/resources/extracted/web/js/dynmaputils.js b/DynmapCore/src/main/resources/extracted/web/js/dynmaputils.js
index 8884a422..63da83e5 100644
--- a/DynmapCore/src/main/resources/extracted/web/js/dynmaputils.js
+++ b/DynmapCore/src/main/resources/extracted/web/js/dynmaputils.js
@@ -207,13 +207,18 @@ var DynmapTileLayer = L.TileLayer.extend({
 	},
 
 	_tickLoadQueue: function() {
-		if (this._loadingTiles.length > 4) {
+		if (this._loadingTiles.length > 12) {
 			return;
 		}
 
+		if (this._loadingTiles.length !== 0) {
+			$(dynmap).trigger("load-from-center");
+		}
+
 		var next = this._loadQueue.shift();
 
 		if (!next) {
+			$(dynmap).trigger("tile-queue-loaded");
 			return;
 		}
 
diff --git a/DynmapCore/src/main/resources/extracted/web/js/map.js b/DynmapCore/src/main/resources/extracted/web/js/map.js
index 0bbcd9f7..8c426376 100644
--- a/DynmapCore/src/main/resources/extracted/web/js/map.js
+++ b/DynmapCore/src/main/resources/extracted/web/js/map.js
@@ -647,7 +647,6 @@ DynMap.prototype = {
 					me.lasttimestamp = update.timestamp;
 				}
 				if(me.options.confighash != update.confighash) {
-				    window.location = me.getLink();
 					return;
 				}
 				me.playerfield.text(me.options['msg-players'] + " [" + update.currentcount + "/" + me.options.maxcount + "]");
